src/proxy/cpsubsystem/SessionlessSemaphoreProxy.ts

```diff
@@ -0,0 +1,184 @@
+import {SemaphoreAcquireCodec} from '../../codec/SemaphoreAcquireCodec';
+import {SemaphoreAvailablePermitsCodec} from '../../codec/SemaphoreAvailablePermitsCodec';
+import {SemaphoreDrainCodec} from '../../codec/SemaphoreDrainCodec';
+import {SemaphoreChangeCodec} from '../../codec/SemaphoreChangeCodec';
+import {SemaphoreReleaseCodec} from '../../codec/SemaphoreReleaseCodec';
+import {
+    IllegalStateError,
+    WaitKeyCancelledError
+} from '../../core';
+
+/** @internal */
+export class SessionlessSemaphoreProxy extends BaseCPProxy implements ISemaphore {
+
+    private readonly sessionManager: CPSessionManager;
+
+    constructor(client: HazelcastClient,
+                groupId: RaftGroupId,
+                proxyName: string,
+                objectName: string) {
+        super(client, CPProxyManager.SEMAPHORE_SERVICE, groupId, proxyName, objectName);
+        this.sessionManager = (client.getCPSubsystem() as CPSubsystemImpl).getCPSessionManager();
+    }
+
+    init(permits: number): Promise<boolean> {
+        assertNonNegativeNumber(permits);
+        return this.encodeInvokeOnRandomTarget(SemaphoreInitCodec, this.groupId, this.objectName, permits)
+            .then((clientMessage) => {
+                const response = SemaphoreInitCodec.decodeResponse(clientMessage);
+                return response.response;
+            });
+    }
+
+    acquire(permits = 1): Promise<void> {
+        assertNonNegativeNumber(permits);
+
+        return this.doTryAcquire(permits, -1).then();
+    }
+
+    tryAcquire(permits: number, timeout = 0): Promise<boolean> {
+        assertNonNegativeNumber(permits);
+        assertNonNegativeNumber(timeout);
+
+        return this.doTryAcquire(permits, timeout);
+    }
+
+    private doTryAcquire(permits: number, timeout: number): Promise<boolean> {
+        const invocationUid = UuidUtil.generate();
+        return this.getClusterWideThreadId()
+            .then((clusterWideThreadId) =>
+                this.encodeInvokeOnRandomTarget(
+                    SemaphoreAcquireCodec,
+                    this.groupId,
+                    this.objectName,
+                    NO_SESSION_ID,
+                    clusterWideThreadId,
+                    invocationUid,
+                    permits,
+                    Long.fromNumber(timeout)
+                )
+            )
+            .then((clientMessage) => {
+                const response = SemaphoreAcquireCodec.decodeResponse(clientMessage);
+                return response.response;
+            })
+            .catch((err) => {
+                if (err instanceof WaitKeyCancelledError) {
+                    throw new IllegalStateError('Semaphore[' + this.objectName
+                        + '] not acquired because the acquire call on the CP group was cancelled.');
+                }
+                throw err;
+            });
+    }
+
+    release(permits = 1): Promise<void> {
+        assertNonNegativeNumber(permits);
```

Comment:
Java client expects permits to be positive. But it seems that we are accepting 0 too. I think we should change that

Reply:
Addressed in ee862f17b4d5fbbf4c12430c29d3045a893ffedf